File: String/28-Find-the-Index-of-the-First-Occurrence-in-a-String.java
class Solution {
    public int strStr(String haystack, String needle) {
        int lenHaystack = haystack.length();
        int lenNeedle = needle.length();

        if (lenHaystack < lenNeedle) {
            return -1;
        }

        // Important: the i needs be to less or equal to the lenHaystack - lenNeedle.
        for (int i = 0; i <= lenHaystack - lenNeedle; i++) {
            int j = 0;
            while (j < lenNeedle && haystack.charAt(i + j) == needle.charAt(j)) {
                j++;
            }

            if (j == lenNeedle) {
                return i;
            }
        }
        return -1;
    }
}

File: String/28-Find-the-Index-of-the-First-Occurrence-in-a-String.ts
function strStr(haystack: string, needle: string): number {
    let lenHaystack = haystack.length;
    let lenNeedle = needle.length;

    if (lenHaystack < lenNeedle) {
        return -1;
    }

    for (let i = 0; i <= lenHaystack - lenNeedle; i++) {
        let j = 0;

        while (j < lenNeedle && haystack[i + j] == needle[j]) {
            j++;
        }

        if (j == lenNeedle) {
            return i;
        }
    }

    return -1;
};